### PR TITLE
qemu: Upgrade qemu 9.1.2

### PR DESCRIPTION
**Pull request by zvonkok**
New features added in the major release have been disabled for now.
Deprecated features related to live-block-migration and avx512f have been removed.
Plan is to test the newly added features for performance and add them as
required in the future.

The major release includes quite a few bug fixes and improvments related
to storage and memory:

Block improvements:

The block layer is now thread safe and multiple I/O threads can send I/O
requests to the same disk image. virtio-blk can be configured to make use
of this and improve scalability
Fixed I/O errors in the libblkio based block drivers caused by incorrect
memory alignment of bounce buffers
In 'blklogwrites', potential log corruption bugs were fixed related to
concurrent requests and to write zeroes operations
Memory improments:

Memory backends specified on the QEMU commandline can now be preallocated
concurrently/asynchronously in some configurations, to speedup QEMU startup
with preallocation on multiple memory backends: one requirement is that memory
backends that are getting preallocated have a "thread-context" set.
QEMU 8.2 accidentally allowed for creation of memory backends with sizes that
are not aligned to the (huge) page size. This has been fixed.

### CODE REVIEW

**Review by fidencio — APPROVED** [review]
lgtm, thanks @zvonkok!

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (stevenhorsman) [inline tools/packaging/scripts/configure-hypervisor.sh:432]: @zvonkok - does this effectively remove 9p support? On the AC call the impression I got was that it was just pmem support that was blocking this PR, so we could do the 9p removal in a less rushed fashion. Did I misunderstand?

Comment (fidencio) [reply tools/packaging/scripts/configure-hypervisor.sh:432]: Steve, you raised a very good point here.

I think 9p support from the runtime / guest kernel should be dropped together with this change, most likely on its own PR.

**Comment by fidencio** [comment]
@zvonkok, let me put this one as a draft till we solve the 9p situation, and avoid it getting mistakenly merged.

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (stevenhorsman) [reply tools/packaging/scripts/configure-hypervisor.sh:438]: @zvonkok - I'm not sure why this has been resolved, as we agreed to keep 9p until post-3.12?

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (stevenhorsman) [inline tools/packaging/scripts/configure-hypervisor.sh:436]: Ok, the change was reverted in a different commit, so I think this bit needs to go into the earlier one that removed it.

File: versions.yaml
Comment (gkurz) [inline versions.yaml:92]: @zvonkok So, this series first bumps to 9.1.0 and then it bumps again to 9.1.1 here. Why not bumping QEMU just once in the same commit ?

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (gkurz) [inline tools/packaging/scripts/configure-hypervisor.sh:355]: How does this relate to pmem ?

File: tools/packaging/static-build/qemu/Dockerfile
Comment (gkurz) [inline tools/packaging/static-build/qemu/Dockerfile:26]: @zvonkok - same comment here, I hardly understand how bumping the ubuntu version relates to the disabling pmem...

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (gkurz) [inline tools/packaging/scripts/configure-hypervisor.sh:426]: @zvonkok - it looks like this belongs to some other patch...

File: tools/packaging/static-build/qemu/Dockerfile
Comment (gkurz) [inline tools/packaging/static-build/qemu/Dockerfile:78]: @zvonkok - I'm assuming this change is because a liburing pkg is available in jammy, right ? If so, please move that to a separate commit.

File: tools/packaging/static-build/qemu/Dockerfile
Comment (gkurz) [inline tools/packaging/static-build/qemu/Dockerfile:56]: @zvonkok - it looks this should go along with the ubuntu bump.

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (zvonkok) [reply tools/packaging/scripts/configure-hypervisor.sh:438]: Ah shoot. I though I removed that.

File: tools/packaging/static-build/qemu/Dockerfile
Comment (zvonkok) [reply tools/packaging/static-build/qemu/Dockerfile:26]: For building 9.1.1 we need a new package, namely python-tomli which is only available in jammy.

File: tools/packaging/static-build/qemu/Dockerfile
Comment (zvonkok) [reply tools/packaging/static-build/qemu/Dockerfile:56]: You mean in another commit?

File: tools/packaging/static-build/qemu/Dockerfile
Comment (zvonkok) [reply tools/packaging/static-build/qemu/Dockerfile:78]: Alrighty, yeah this commmit grew hysterically.

File: versions.yaml
Comment (zvonkok) [reply versions.yaml:92]: We're bumping it to 9.1.1 where do you see a reference to 9.1.0?

Comment (gkurz) [reply versions.yaml:92]: > We're bumping it to 9.1.1 where do you see a reference to 9.1.0?

The first commit  bumps from 8.2.4 to 9.1.0 at https://github.com/kata-containers/kata-containers/pull/10492/commits/521dab29550839fd734fa782e2764980aa857bc1#diff-1c4cd801df522f4a92edbfb0fea95364ed074a391ea47c284ddc078f512f7b6aL91-L92

File: tools/packaging/static-build/qemu/Dockerfile
Comment (gkurz) [reply tools/packaging/static-build/qemu/Dockerfile:78]: > Alrighty, yeah this commmit grew hysterically.

It is isn't just about that. The changes around liburing should appear in the same patch.

File: tools/packaging/static-build/qemu/Dockerfile
Comment (gkurz) [reply tools/packaging/static-build/qemu/Dockerfile:56]: Yes, this should go along with the removal of the manual build of liburing, either in the commit that bumps ubuntu or a subsequent commit.

File: versions.yaml
Comment (zvonkok) [reply versions.yaml:92]: Ah this was a leftover from the previous PR let me rectify this.

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (zvonkok) [reply tools/packaging/scripts/configure-hypervisor.sh:425]: We need io-uring support building without explicitly stating it fails to build 9.1.1 somehow.

**Comment by Apokleos** [comment]
@zvonkok I have a question that, I notice that the qemu's latest release is [v9.1.2](https://www.qemu.org/), but we now upgrade it up to `qemu-v9.1.1`. Is there some reason not to directly upgrade it to v9.1.2 ?

**Comment by gkurz** [comment]
> @zvonkok I have a question that, I notice that the qemu's latest release is [v9.1.2](https://www.qemu.org/), but we now upgrade it up to `qemu-v9.1.1`. Is there some reason not to directly upgrade it to v9.1.2 ? cc @gkurz

I'd recommend to always bump to the latest fix version when possible.

**Review by gkurz — COMMENTED** [review]
Hi @zvonkok !

This looks much better. I've posted some suggestions.

And then, I have an overall concern on the patch ordering which isn't `git bisect` friendly. It is generally considered better to have a bisectable tree (it is even mandatory for some softwares like the linux kernel or QEMU).

Ideally we'd want a static build of QEMU to succeed for every commits in the PR. This isn't the case here since the QEMU version is bumped first and the build is fixed by the next commits.

It certainly makes sense to use this order while developing but at submission time, it is better to have the fixes or preparatory commits to go first and only then enable the new version. It is also convenient to squash the deps that are directly related to the version bump into the same commit.

For this PR, this would basically mean :

1. qemu: Remove abs path
2. qemu: Build as user
3. qemu: Use proper QEMU builder
4. qemu: Bump Ubuntu version in Dockerfile + qemu: Add depedency
5. qemu: Replace from source build with package
6. qemu: Disable pmem
7. qemu: Bump version + qemu: Add new options for 9.1.2 + qemu: Add no_patches.txt

WDYT @zvonkok ?

**Comment by gkurz** [comment]
Hi @zvonkok ! Thanks for reworking the PR. We're nearly there :smiley:. Just a couple of nits :

The description of commit `qemu: Disable pmem` contains a binary character :
```
We're disabling pmem support, it is heavilly broken with
Ubuntu's static build of QEMU￼and not needed
```
and the fix requested in https://github.com/kata-containers/kata-containers/pull/10492/commits/3f2b5e38bd2b33e7dd15cd4031e40564a31ce230#r1873495320 was put in the _wrong_ commit.

Rest looks good to me. I'll approve when this is fixed.

File: tools/packaging/scripts/configure-hypervisor.sh
Comment (stevenhorsman) [inline tools/packaging/scripts/configure-hypervisor.sh]: Should we remove `qemu_options+=(functionality:--enable-libpmem)` as we are calling `qemu_options+=(functionality:--disable-libpmem)` right after, which is a bit confusing?

Comment (stevenhorsman) [reply tools/packaging/scripts/configure-hypervisor.sh]: Ah, this is done in a later commit. I guess that implicitly seconds Greg suggestion to move it :)

**Review by gkurz — APPROVED** [review]
/lgtm

Thanks @zvonkok !

**Review by fidencio — APPROVED** [review]
still lgtm, thanks @zvonkok (and @gkurz for the ping).

**Review by stevenhorsman — APPROVED** [review]
LGTM. Thanks for the changes @zvonkok and great suggestions @gkurz!

**Pull request merged**
